### PR TITLE
UCT/IB/DC: Code simplifications

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

DC code improvements.

## Why ?

To reduce code duplications and decrease amount of code.

## How ?

1. Remove separate `if` block for randmo DCI policy to do arbiter group purge in `uct_dc_mlx5_ep_pending_purge`.
2. Remvoe branch `uct_dc_mlx5_iface_is_dci_rand` from `uct_dc_mlx5_iface_dci_detach`.

### CODE REVIEW

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 1]
```diff
@@ -1481,30 +1481,20 @@ void uct_dc_mlx5_ep_pending_purge(uct_ep_h tl_ep,
 {
     uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_mlx5_iface_t);
     uct_dc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
+    uint8_t dci                = ep->dci;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1484]: maybe remove this var?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 2]
```diff
@@ -1481,30 +1481,20 @@ void uct_dc_mlx5_ep_pending_purge(uct_ep_h tl_ep,
...
     ucs_arbiter_group_purge(waitq, group, uct_dc_mlx5_ep_arbiter_purge_cb,
                             &args);
 
-    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
-        return;
-    }
-
-    dci = ep->dci;
-    if (!uct_dc_mlx5_iface_dci_detach(iface, ep)) {
+    if ((dci == UCT_DC_MLX5_EP_NO_DCI) ||
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1495]: maybe
if (dci != NO_DCI && !is_rand && detach) { schedule_release }
and similar refactor (+remove dci helper var) in uct_dc_mlx5_iface_dci_do_dcs_pending_tx
BTW maybe we can move schedule_release inside detach ?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 3]
```diff
@@ -1481,30 +1481,20 @@ void uct_dc_mlx5_ep_pending_purge(uct_ep_h tl_ep,
...
     ucs_arbiter_group_t *group;
     uint8_t pool_index;
-    uint8_t dci;
-
-    if (uct_dc_mlx5_iface_is_dci_rand(iface)) {
-        ucs_arbiter_group_purge(uct_dc_mlx5_iface_tx_waitq(iface),
-                                uct_dc_mlx5_ep_rand_arb_group(iface, ep),
-                                uct_dc_mlx5_ep_arbiter_purge_cb, &args);
-        return;
-    }
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5_ep.c:1496]: in this case we may deschedule group for rand scheme on line 1511, which is not allowed

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 1 above)
Comment (dmitrygx) [reply src/uct/ib/dc/dc_mlx5_ep.c]: we do `uct_dc_mlx5_iface_dci_detach`, so we should save `dci`

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 3 above)
Comment (dmitrygx) [reply src/uct/ib/dc/dc_mlx5_ep.c:1496]: we have a check for `uct_dc_mlx5_iface_is_dci_rand` below to return from the function

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 2 above)
Comment (dmitrygx) [reply src/uct/ib/dc/dc_mlx5_ep.c]: good idea, done

**Review by yosefe — DISMISSED** [review]

**Review by brminich — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Pull request merged**